### PR TITLE
Feature/switch between accounts

### LINKED ISSUE #32: As a signed in user I can switch between accounts
If a user is a collaborator in another user's account, we want to allow them to switch to that account via a dropdown in the header. The dropdown should list all the accounts for which I am a collaborator and my default account as well. Once the user switches to an account, it will only have access to the dashboard and not to any other pages for the current account, e.g (collaborators, clients, reports).

- Add dropdown for choosing an account
- Add Pundit for authorization

### PR DESCRIPTION
**Pull request by philip13**
closes #32

### CODE REVIEW

File: app/controllers/application_controller.rb [hunk 1]
```diff
@@ -1,5 +1,12 @@
 class ApplicationController < ActionController::Base
   include SetupDeviseBehavior
+  include Pundit::Authorization
 
-  before_action :authenticate_user!
+  rescue_from Pundit::NotAuthorizedError, with: :user_not_authorized
+
+  private
+    def user_not_authorized
+      flash[:alert] = "You are not authorized to perform this action."
```
Comment (ingilniero) [inline app/controllers/application_controller.rb:9]: Can you add a translation for this?

File: app/models/user.rb [hunk 2]
```diff
@@ -6,4 +6,8 @@ class User < ApplicationRecord
   devise :database_authenticatable, :registerable, :recoverable, :rememberable, :validatable, :invitable
 
   accepts_nested_attributes_for :account
+
+  def is_collaborator?
```
Comment (ingilniero) [inline app/models/user.rb:10]: remove the `is` prefix

File: app/views/shared/_header.html.erb [hunk 3]
```diff
@@ -11,44 +11,36 @@
 
       <input class="w-32 pl-10 pr-4 text-indigo-600 border-gray-200 rounded-md sm:w-64 focus:border-indigo-600 focus:ring focus:ring-opacity-40 focus:ring-indigo-500" type="text" placeholder="<%= t(".search_placeholder") %>" />
     </div>
+    <div class="relative mx-4 px-2 lg:mx-0">
+      <span class="font-semibold">
+        Workspace:
```
Comment (ingilniero) [inline app/views/shared/_header.html.erb:16]: Add translation for this

File: test/policies/account_policy_test.rb [hunk 4]
```diff
@@ -0,0 +1,18 @@
+require "test_helper"
+
+class AccountPolicyTest < ActiveSupport::TestCase
```
Comment (ingilniero) [inline test/policies/account_policy_test.rb:3]: remove this file

File: app/controllers/accounts_controller.rb [hunk 5]
```diff
@@ -1,5 +1,8 @@
 class AccountsController < ApplicationController
   def show
-    @account = current_user.account
+    @account = params[:id] ? Account.find(params[:id]) : current_user.account
```
Comment (ingilniero) [inline app/controllers/accounts_controller.rb:3]: Add a rescue clause for not found accounts.

File: app/controllers/accounts_controller.rb [hunk 6]
```diff
@@ -1,5 +1,8 @@
...
+    authorize @account, :show?
+    session[:account_id] = @account.id
```
Comment (ingilniero) [inline app/controllers/accounts_controller.rb:5]: I think you can use https://api.rubyonrails.org/classes/ActiveSupport/CurrentAttributes.html to make the access to the current account and current user's account more accessible. e.g:

```
Current.current_user_account
Current.current_account
```

File: app/helpers/current_account_helper.rb [hunk 7]
```diff
@@ -0,0 +1,5 @@
+module CurrentAccountHelper
+  def trunc_account_name account_name
+    account_name.truncate(15, separator: " ")
```
Comment (ingilniero) [inline app/helpers/current_account_helper.rb:3]: move this to the `Account` model, maybe something like `small_name`.

File: app/policies/account_policy.rb [hunk 8]
```diff
@@ -0,0 +1,18 @@
...
+
+  def initialize(user, account)
+    raise Pundit::NotAuthorizedError, "must be a collaborator" unless user
+    @user = user
+    @account = account
+    @my_account = @user.account
+  end
+
+  def show?
+    @my_account == @account || @my_account.get_accounts.include?(@account)
```
Comment (ingilniero) [inline app/policies/account_policy.rb:12]: Maybe the `get_accounts` should be in the user.

File: app/views/shared/_header.html.erb [hunk 9]
```diff
@@ -11,44 +11,36 @@
...
-          <a href="#" class="block px-4 py-2 text-sm text-gray-700 hover:bg-indigo-600 hover:text-white">Products</a>
-          <router-link to="/" class="block px-4 py-2 text-sm text-gray-700 hover:bg-indigo-600 hover:text-white">Log out</router-link>
+      <transition enter-active-class="transition duration-150 ease-out transform" enter-from-class="scale-95 opacity-0" enter-to-class="scale-100 opacity-100" leave-active-class="transition duration-150 ease-in transform" leave-from-class="scale-100 opacity-100" leave-to-class="scale-95 opacity-0">
+        <div class="absolute right-0 z-20 w-48 py-2 mt-2 bg-white rounded-md shadow-xl hidden" data-menu-target="options">
+            <a href="#" class="block px-4 py-2 text-sm text-gray-700 hover:bg-indigo-600 hover:text-white">Profile</a>
+            <% if current_user.is_collaborator? %>
+              <% user_accounts = current_user.account.get_accounts %>
+              <% user_accounts.each do |a| %>
+                <%= link_to a.name.truncate(15, separator: ' '), account_path(a), class: "block px-4 py-2 text-sm text-gray-700 hover:bg-indigo-600 hover:text-white"  %>
```
Comment (ingilniero) [inline app/views/shared/_header.html.erb:40]: one you have the `small_name` method in the account you can use it here.

File: app/views/shared/_header.html.erb [hunk 10]
```diff
@@ -11,44 +11,36 @@
...
+      <div class="fixed inset-0 z-10 w-full h-full hidden" data-menu-target="clearfix"></div>
 
-          >
-          <a href="#" class="block px-4 py-2 text-sm text-gray-700 hover:bg-indigo-600 hover:text-white">Profile</a>
-          <a href="#" class="block px-4 py-2 text-sm text-gray-700 hover:bg-indigo-600 hover:text-white">Products</a>
-          <router-link to="/" class="block px-4 py-2 text-sm text-gray-700 hover:bg-indigo-600 hover:text-white">Log out</router-link>
+      <transition enter-active-class="transition duration-150 ease-out transform" enter-from-class="scale-95 opacity-0" enter-to-class="scale-100 opacity-100" leave-active-class="transition duration-150 ease-in transform" leave-from-class="scale-100 opacity-100" leave-to-class="scale-95 opacity-0">
+        <div class="absolute right-0 z-20 w-48 py-2 mt-2 bg-white rounded-md shadow-xl hidden" data-menu-target="options">
+            <a href="#" class="block px-4 py-2 text-sm text-gray-700 hover:bg-indigo-600 hover:text-white">Profile</a>
+            <% if current_user.is_collaborator? %>
```
Comment (ingilniero) [inline app/views/shared/_header.html.erb:37]: maybe we should use another dropdown for the accounts or at least add a seperator in this one.

File: app/views/shared/_header.html.erb (hunk 9 above)
Comment (ingilniero) [inline app/views/shared/_header.html.erb:40]: add the "active" classes for the active account.

**Review by ingilniero — CHANGES_REQUESTED** [review]

File: app/controllers/application_controller.rb (hunk 1 above)
Comment (philip13) [reply app/controllers/application_controller.rb:9]: My Idea is to create a file like this:
`config/locales/policy.xx.yml`

any suggestion?

Comment (ingilniero) [reply app/controllers/application_controller.rb:9]: you can put it in globals

File: app/policies/account_policy.rb (hunk 8 above)
Comment (philip13) [reply app/policies/account_policy.rb:12]: We can talk tomorrow, to listen your suggestion

File: app/models/account.rb [hunk 11]
```diff
@@ -3,4 +3,12 @@ class Account < ApplicationRecord
 
   validates :name, presence: true
   validates :name, uniqueness: true
+
+  def get_accounts
+    Account.where("id IN (?)", [id, owner.invited_by.account.id])
```
Comment (ingilniero) [inline app/models/account.rb:8]: you can rewrite this as `Account.where(id: [id, owner.invited_by.account.id])`

**Review by ingilniero — CHANGES_REQUESTED** [review]

File: app/controllers/accounts_controller.rb [hunk 12]
```diff
@@ -1,5 +1,13 @@
 class AccountsController < ApplicationController
   def show
-    @account = current_user.account
+    begin
+      @account = params[:id] ? Account.find(params[:id]) : current_user.account
+      authorize @account, :show?
+    rescue
+      @account = current_user.account
+      redirect_to "/404"
```
[the remark this answers is not in the document]
Comment (ingilniero) [reply app/controllers/accounts_controller.rb:8]: You can use `rescue_from`

**Review by ingilniero — APPROVED** [review]

File: app/views/shared/_header.html.erb [hunk 13]
```diff
@@ -37,7 +37,7 @@
               Profile
             </a>
             <% if current_user.collaborator? %>
-              <% user_accounts = current_user.account.get_accounts %>
+              <% user_accounts = current_user.get_accounts %>
```
[the remark this answers is not in the document]
Comment (philip13) [reply app/views/shared/_header.html.erb:40]: That means,  change to `Current.user.collaborator?` ... ?

[the remark this answers is not in the document]
Comment (ingilniero) [reply app/views/shared/_header.html.erb:40]: yes, the `user_accounts` variable

**Pull request merged**